onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu

```diff
@@ -805,6 +806,189 @@ void UpdateDecoderMaskedMultiHeadAttentionCacheIndirection(int32_t* tgt_indir_ca
                                                                                           max_seq_length,
                                                                                           current_length);
 }
+
+#ifndef USE_ROCM
+namespace {
+template <typename T, size_t size>
+struct TypeMapper : public V_vec_m_<T, size> {};
+
+template <>
+struct TypeMapper<int32_t, 2> {
+  using Type = uint2;
+};
+
+template <>
+struct TypeMapper<int32_t, 4> {
+  using Type = uint4;
+};
+}  // namespace
+#endif
+
+template <typename T>
+__global__ void KeyCacheExpansionKernel(const T* input,
+                                        T* output,
+                                        int beam_width,
+                                        int max_seq_length,
+                                        int head_size) {
+  const int num_heads = gridDim.y;
+  const int sequence_length = gridDim.z;
+
+  const int bbid = blockIdx.x;
+  const int batch_id = bbid / beam_width;
+  const int head_id = blockIdx.y;
+  const int s = blockIdx.z;
+  const int tidx = threadIdx.x;
+
+  const int input_offset = ((batch_id * num_heads + head_id) * sequence_length + s) * head_size + tidx;
+  const int output_offset = ((bbid * num_heads + head_id) * max_seq_length + s) * head_size + tidx;
+
+  if (tidx < head_size) {
+    output[output_offset] = input[input_offset];
+  }
+}
+
+template <typename T>
+void KeyCacheExpansionKernelLauncher(const T* key_cache,
+                                     T* key_cache_expanded,
+                                     int batch_size,
+                                     int beam_width,
+                                     int num_heads,
+                                     int sequence_length,
+                                     int max_seq_length,
+                                     int head_size,
+                                     cudaStream_t stream) {
+  const dim3 block(256);
```

Comment:
>block(256); [](http://example.com/codeflow?start=13&length=11)

block size is too large. head_size is usually 32 or 64.

Reply:
we can use nearest 2^^n